### ISSUE TITLE
Some water in Trails through Daybreak renders incorrectly on Nvidia GPU's

### ISSUE THREAD

**Issue by UltraSunDK** [doitsujin/dxvk#4162]
Running The Legend of Heroes: Trails through Daybreak with DXVK on Windows and reportedly Linux machines using Nvidia GPU's  results in water rendering at Thaarbad (ingame location) to render incorrectly.

The issue shows itself by odd looking lines in the water texture that are not present normally otherwise:
Image of issue:
![20240720233539_1](https://github.com/user-attachments/assets/a910a736-bb8e-4261-937d-c23ea4667c72)

Here is how it's supposed to look:
![20240720233717_1](https://github.com/user-attachments/assets/a80542b5-648d-436a-9e0f-4d37bf8592aa)

I do not currently have a Nvidia powered machine running Linux on hand, but a user on the Steam forums for Trails through Daybreak reported their Linux Mint 21.3 PC with a Nvidia GeForce RTX 3070ti and Intel i9-9900kf have the issue and I on Windows using Nvidia hardware can reproduce it too using the latest version of DXVK. Forum thread in question is linked here: https://steamcommunity.com/app/2138610/discussions/0/4521134886148456576/

I have additionally been able to confirm with others that the issue is not present on the Steam Deck nor on Windows using DXVK on AMD GPU's. This therefore appears to be an exclusively Nvidia issue on both Windows and Linux when using DXVK in this game. 

A save game at the location is linked here, for if you own the game to check with it: https://1drv.ms/u/s!AsdkXaek-4Dwj8x9UOQ3ngM_mAWJQg?e=URGx6G

### Software information
Game: The Legend of Heroes: Trails through Daybreak
Platform: Steam
Game version: v1.3.5

I have been able to confirm the issue is present on whatever settings is used, including both Water Shading options.
The only necessity is a Nvidia GPU.

### System information
Windows PC:
- GPU: Nvidia GeForce RTX 4090
- Driver: 560.70
- Wine version: N/A
- DXVK version: 2.4

### Apitrace file(s)
Apitrace file:
https://1drv.ms/u/s!AsdkXaek-4Dwj8x-rVF2mNIiZEe3jQ?e=q12gyI
Note: This was generated using the quick load feature that Trails through Daybreak features.

### Log files
DXVK logs from Windows
[kuro_dxgi.log](https://github.com/user-attachments/files/16322450/kuro_dxgi.log)
[kuro_d3d11.log](https://github.com/user-attachments/files/16322451/kuro_d3d11.log)

**Comment by Blisto91** [comment]
Thank you for a well made issue.
I can see that the fog just above the horizon doesn't render in either with dxvk. I assume that is also a bug and not a settings change?

**Comment by UltraSunDK** [comment]
> Thank you for a well made issue. I can see that the fog just above the horizon doesn't render in either with dxvk. I assume that is also a bug and not a settings change?

It's not a settings change indeed and something that was noticed after submission on my end. Both screenshots should have been taken using the same settings. The fog in that background is slightly moving however, but after multiple loads it did seem consistent in its difference.

If you need any more info on that matter (including more pictures/api traces) for other settings combinations, I'll be happy to provide it.

**Comment by K0bin** [comment]
Does disabling MSAA (if possible) make a difference?

**Comment by UltraSunDK** [comment]
> Does disabling MSAA (if possible) make a difference?
The game supports the following AA options natively:

- Off
- Basic AA - Blur filter
- FXAA
- MSAA - 2x to 8x
- SGSSAA 2x to 8x
- Transparency Multisampling 

In my testing on Windows the issue is present with every AA setting including Off.
I have also attempted to lower every setting to their absolute minimum, with no changes in the reported water issue when using DXVK.

**Comment by doitsujin** [comment]
Shader seems to have some fragile math involving trigonometric functions going on, which seems to be the cause of the issue but also makes things hard to debug since differences in output between different drivers are entirely expected and within spec.

TL;DR no idea what's going on, I don't see any obvious bug on our end though.

**Comment by PeterTh** [comment]
The water shader indeed relies on specific outputs in some rather insignificant bits of trigonometric functions for its perlin noise implementation. For Kuro 2, we actually had to replace this noise function because it would break (natively in D3D11) specifically on AMD hardware due to floating point inaccuracy (which is rather amusing, given that it was written for PS5 originally).

@doitsujin if you think it helps I can provide debug shaders and/or shader code for this, but I could believe that it's relying on unspecified behaviour.

**Comment by UltraSunDK** [comment]
I've retested the scene with the referenced pull request above and it appears to fix the issue on Windows.
![20240808173549_1](https://github.com/user-attachments/assets/e8347433-a825-4baa-a97a-de81adfe9e38)

I will however have to note one thing I forgot to mention in the initial report. There are actually two different PC versions of Trails through Daybreak from two different developers. The pull request fixes the Western/Japanese version developed by PH3 games, but the Chinese/Korean version published by Clouded Leopard Entertainment has the same issue with the water, but as they use a different name for the executable, the fix is not applied for that version.

Here is a image of the Clouded Leopard Entertainment version.
![20240808173044_1](https://github.com/user-attachments/assets/50b2b9d7-f377-431b-8792-12f8e33c8b2c)

Here is Clouded Leopard Entertainment with the executable renamed from ed9.exe to kuro.exe
![image](https://github.com/user-attachments/assets/b5819281-a800-40d5-8df5-ca44e1b35005)

So the pull request fixes both versions of the game, but as Clouded Leopard Entertainment named the executable ed9.exe the change is not applied automatically to that version of the game.

**Comment by doitsujin** [comment]
Added ed9.exe too for now. Applying this to kuro 2 won't be an issue since we're planning to make this the default behaviour anyway relatively soon.
